File: Sorting/Medium/Shell Sort/shellsort.py
'''Copyrights to venkys.io
For more information, visite https://venkys.io"/

Python program to sort a given array in order using Shell sort algorithm.'''

# Stable : No
# Inplace : Yes
# Adaptive : Yes

# Space complexity: O(1)

#Time Complexity:O(n^2) in the worst-case scenario, and O(n log n) in the average case scenario.

def VSDShellSort(arr,n):
    shellsize=8  # Initialize the initial gap size for Shell Sort
    while(shellsize>=1):# Continue the loop until the gap size becomes 1
        i=shellsize
        k=0
        while(i<n and k<shellsize): # Iterate through the array with the current gap size
            key=arr[i] # Store the current element as the key
            j=i-shellsize  # Start comparing with the element 'shellsize' positions before the current element
            while j>=0:
                if(arr[j]>key):
                    arr[j+shellsize]=arr[j] # Shift elements greater than the key to the right
                else:
                    break
                j-=shellsize
            arr[j+shellsize]=key # Insert the key at its correct position in the sorted sequence
            if(i+shellsize)>=n:
                k+=1
                i=k+shellsize # Move to the next group of elements for comparison
            else:
                i+=shellsize
        shellsize=shellsize//2 # Reduce the gap size by half in each iteration
    print(*arr,sep=" ")  # Print the sorted array


def main():
    try:
        # Input the number of elements
        n = int(input("Enter the number of elements: "))
        # Check if the entered number of elements is not a positive integer
        if n <= 0:
            raise ValueError("Number of elements should be a positive integer.")

        # Input the elements of the array and convert them to integers
        arr = list(map(int, input("Enter the elements separated by space: ").split()))

        # Check if the number of elements entered matches the specified count
        if len(arr) != n:
            raise ValueError("Number of elements entered doesn't match the specified count.")

        # Call the VSDShellSort function to sort the array
        VSDShellSort(arr, n)

    except ValueError as e:
        # Handle the case where an error occurs (e.g., non-integer input)
        print(f"Error: {e}")

# Check if the script is executed as the main program
if __name__ == "__main__":
    main()

File: Sorting/Medium/Shell Sort/ShellSort.java
/* Copyrights to venkys.io */
/* For more programs, visit venkys.io */
/* Java program for Shell sort */

// Stable: No
// Inplace: Yes
// Adaptive: Yes

// Space complexity: O(1)

// Time Complexity: O(n^2) in the worst-case scenario, and O(n log n) in the average case scenario.

import java.util.Scanner; // Importing Scanner class from java.util package for user input

public class ShellSort {
    static int n; // Size of the array
    static int[] arr; // Array of numbers to be sorted

    // Function to perform shell sort
    public static void shellSort() {
        int shellSize = 8; // Initial shell size
        
        // Continue the sorting process until the shell size becomes 0
        while (shellSize >= 1) {
            // Iterate through the array with the current shell size
            for (int i = shellSize, k = 0; i < n && k < shellSize;) {
                int key = arr[i];// Store the current element to be inserted
                int j = i - shellSize; 
                
                // Move elements greater than key to the right until a smaller or equal element is found,
                // or until the beginning of the subarray is reached.
                for (; j >= 0; j -= shellSize) {
                    if (arr[j] > key) {
                        arr[j + shellSize] = arr[j];// Shift the current element to the right
                    } else {
                        break; // Stop shifting if a smaller or equal element is found
                    }
                }
                arr[j + shellSize] = key;// Insert the stored element at its correct position
                
                // Move to the next position in the array considering the shell size
                if (i + shellSize >= n) {
                    k += 1;
                    i = k + shellSize;
                } else {
                    i += shellSize;
                }
            }
            
            // Reduce the shell size for the next iteration
            shellSize = shellSize / 2;
        }
    }

    // Function to print output
    public static void printOutput() {
        System.out.println("The sorted order is:");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void main(String args[]) {
        // Create a Scanner object to read user input from the console.
        Scanner scanner = new Scanner(System.in);

        try {
            // Input the elements separated by space
            System.out.print("Enter the elements separated by space: ");
            // Read the entire line as input.
            String input = scanner.nextLine();

            // Check for null or empty input
            if (input == null || input.trim().isEmpty()) {
                System.out.println("Error: Empty input. Please enter valid elements.");
                return; // Exit the program to handle the error.
            }

            // Split the input string into an array of string elements based on space.
            String[] elements = input.split(" ");
            n = elements.length;// Set the size of the array based on the number of elements.
            arr = new int[n];// Initialize the array to store the parsed integers.

            // Parse input elements to integers
            for (int i = 0; i < n; i++) {
                try {
                    // Parse each element as an integer and assign it to the array.
                    arr[i] = Integer.parseInt(elements[i]);
                } catch (NumberFormatException e) {
                    // Handle the case where an element is not a valid integer.
                    System.out.println("Error: Invalid input. Please enter valid integers.");
                    return;// Exit the program to handle the error.
                }
            }

            // Call the shellSort function to perform the sorting.
            shellSort();

            // Call the printOutput function to display the sorted array.
            printOutput();
        } finally {
            // Close the scanner to avoid resource leaks
            if (scanner != null) {
                scanner.close();
            }
        }
    }
}
